rust/lance-index/src/vector/sq.rs

```diff
@@ -0,0 +1,258 @@
+// Copyright 2024 Lance Developers.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::{any::Any, sync::Arc};
+
+use arrow::array::AsArray;
+use arrow_array::{Array, ArrayRef, FixedSizeListArray, UInt8Array};
+
+use lance_arrow::*;
+use lance_core::{Error, Result};
+use lance_linalg::distance::{Dot, MetricType, L2};
+use num_traits::*;
+use snafu::{location, Location};
+
+#[async_trait::async_trait]
+pub trait ScalarQuantizer: Send + Sync + std::fmt::Debug {
+    fn as_any(&self) -> &dyn Any;
+
+    /// Transform a vector column to SQ code column.
+    ///
+    /// Parameters
+    /// ----------
+    /// *data*: vector array, must be a `FixedSizeListArray`
+    ///
+    /// Returns
+    /// -------
+    ///   SQ code column
+    async fn transform(&self, data: &dyn Array) -> Result<ArrayRef>;
+
+    /// Get the centroids for each dimension.
+    fn num_bits(&self) -> u16;
+
+    /// Whether to use residual as input or not.
+    fn use_residual(&self) -> bool;
+}
+
+/// Scalar Quantization, optimized for [Apache Arrow] buffer memory layout.
+///
+//
+// TODO: move this to be pub(crate) once we have a better way to test it.
+#[derive(Debug)]
+pub struct ScalarQuantizerImpl<T: ArrowFloatType + Dot + L2> {
+    /// Number of bits for the centroids.
+    ///
+    /// Only support 8, as one of `u8` byte now.
+    pub num_bits: u16,
+
+    /// Distance type.
+    pub metric_type: MetricType,
+
+    pub lower_bound: T::Native,
+    pub upper_bound: T::Native,
+}
+
+impl<T: ArrowFloatType + Dot + L2> ScalarQuantizerImpl<T> {
+    pub fn new(num_bits: u16, metric_type: MetricType) -> Self {
+        Self {
+            num_bits,
+            metric_type,
+            lower_bound: T::MAX,
+            upper_bound: T::MIN,
+        }
+    }
+
+    pub fn with_bounds(
+        num_bits: u16,
+        metric_type: MetricType,
+        bounds: (T::Native, T::Native),
+    ) -> Self {
+        let mut sq = Self::new(num_bits, metric_type);
+        sq.lower_bound = bounds.0;
+        sq.upper_bound = bounds.1;
+        sq
+    }
+
+    pub fn update_bounds(
+        &mut self,
+        vectors: &FixedSizeListArray,
+    ) -> Result<(T::Native, T::Native)> {
+        let data = vectors
```

Comment:
Does it mean that the training need scan over all data ?

Reply:
we don't have to, trained on samples to get the boundary, for the values out of range, we can simply map them to the lower/upper bound